tools/tpm2_key_import.c

```diff
@@ -0,0 +1,578 @@
+    return_code = RSA_generate_key_ex(rsa, 2048, bne, NULL);
+    if (return_code != 1) {
+        printf("RSA_generate_key_ex failed\n");
+        return 1;
+    }
+    BIGNUM *n = BN_bin2bn(pub_modulus, MAX_RSA_KEY_BYTES, NULL);
+    ssl_RSA_set0_key(rsa, n, NULL, NULL);
+    if (n == NULL) {
+        printf("Failed RSA_set0_key\n");
+        return 1;
+    }
+    // Encrypting
+    return_code = RSA_public_encrypt(MAX_RSA_KEY_BYTES, encoded,
+            ctx->encrypted_protection_seed_data, rsa, RSA_NO_PADDING);
+    if (return_code < 0) {
+        printf("Failed RSA_public_encrypt\n");
+    }
+    RSA_free(rsa);
+    BN_free(bne);
+    return true;
+}
+
+void aes_128_cfb_encrypt_buffers(uint8_t *buffer1, uint16_t buffer1_size,
+        uint8_t *buffer2, uint16_t buffer2_size, uint8_t *encryption_key,
+        uint8_t *encrypted_data) {
+    //AES encryption
+    uint8_t to_encrypt_buffer[max_buffer_size];
+    memcpy(to_encrypt_buffer, buffer1, buffer1_size);
+    memcpy(to_encrypt_buffer + buffer1_size, buffer2, buffer2_size);
+
+    uint8_t iv_in[SYM_KEY_SIZE] = { 0 };
+    AES_KEY aes;
+    AES_set_encrypt_key(encryption_key, SYM_KEY_SIZE * 8, &aes);
+
+    int block;
+    int num = 0;
+    for (block = 0; block < (buffer1_size + buffer2_size) / SYM_KEY_SIZE;
+            block++) {
+        AES_cfb128_encrypt(to_encrypt_buffer + (block * SYM_KEY_SIZE),
+                encrypted_data + (block * SYM_KEY_SIZE),
+                SYM_KEY_SIZE, &aes, iv_in, &num, AES_ENCRYPT);
+    }
+    AES_cfb128_encrypt(to_encrypt_buffer + (block * SYM_KEY_SIZE),
+            encrypted_data + (block * SYM_KEY_SIZE),
+            (buffer1_size + buffer2_size) % SYM_KEY_SIZE, &aes, iv_in, &num,
+            AES_ENCRYPT);
+}
+
+void hmac_outer_integrity(uint8_t *buffer1, uint16_t buffer1_size,
+        uint8_t *buffer2, uint16_t buffer2_size, uint8_t *hmac_key,
+        uint8_t *outer_integrity_hmac) {
+
+    uint8_t to_hmac_buffer[max_buffer_size];
+    memcpy(to_hmac_buffer, buffer1, buffer1_size);
+    memcpy(to_hmac_buffer + buffer1_size, buffer2, buffer2_size);
+    uint32_t size_in = 0;
+    HMAC(EVP_sha256(), hmac_key, SHA256_DIGEST_SIZE, to_hmac_buffer,
+            buffer1_size + buffer2_size, outer_integrity_hmac, &size_in);
+}
+
+void create_random_seed_and_sensitive_enc_key(tpm_import_ctx *ctx) {
+
+    RAND_bytes(ctx->protection_seed_data, SHA256_DIGEST_SIZE); //max tpm digest
+    ctx->enc_sensitive_key.b.size = SYM_KEY_SIZE;
+    RAND_bytes(ctx->enc_sensitive_key.b.buffer, SYM_KEY_SIZE);
+
+}
+
+void calc_sensitive_unique_data(tpm_import_ctx *ctx) {
+
+    uint8_t *concatenated_seed_unique = malloc(
+            SHA256_DIGEST_SIZE + SYM_KEY_SIZE);
+
+    memcpy(concatenated_seed_unique, ctx->protection_seed_data,
+            SHA256_DIGEST_SIZE);
+    memcpy(concatenated_seed_unique + SHA256_DIGEST_SIZE, ctx->input_key_buffer,
+            SYM_KEY_SIZE);
+
+    SHA256(concatenated_seed_unique, SHA256_DIGEST_SIZE + SYM_KEY_SIZE,
```

Comment:
Isn't there SHA256 routines that allow updates, then you wouldn't need to copy these buffers around.